### PR TITLE
pkg/server: use structured, contextual logging

### PR DESCRIPTION
**Pull request by stevekuznetsov**
Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

Part of #1694 
/assign @ncdc 
/cc @davidfestal @sttts

### CODE REVIEW

File: pkg/server/controllers.go [hunk 1]
```diff
@@ -88,14 +93,15 @@ func (s *Server) installClusterRoleAggregationController(ctx context.Context, co
...
-	return s.AddPostStartHook("kcp-start-kube-cluster-role-aggregation-controller", func(hookContext genericapiserver.PostStartHookContext) error {
+	return s.AddPostStartHook(postStartHookName(controllerName), func(hookContext genericapiserver.PostStartHookContext) error {
 		go c.Run(ctx, 5)
 		return nil
 	})
 }
 
 func (s *Server) installKubeNamespaceController(ctx context.Context, config *rest.Config) error {
-	config = rest.AddUserAgent(rest.CopyConfig(config), "kube-namespace-controller")
+	controllerName := "kube-namespace-controller"
```
Comment (stevekuznetsov) [inline pkg/server/controllers.go:103]: TODO: something more generic?

File: pkg/server/handler.go [hunk 2]
```diff
@@ -353,7 +355,7 @@ func WithWildcardIdentity(handler http.Handler) http.Handler {
 
 		updatedReq, err := processResourceIdentity(req, requestInfo)
 		if err != nil {
-			klog.Errorf("WithWildcardIdentity: unable to determine resource from path %s", req.URL.Path)
+			klog.FromContext(req.Context()).WithValues("orpeation", "WithWildcardIdentity", "path", req.URL.Path).Error(err, "unable to determine resource")
```
Comment (ncdc) [inline pkg/server/handler.go:358]:
```suggestion
			klog.FromContext(req.Context()).WithValues("operation", "WithWildcardIdentity", "path", req.URL.Path).Error(err, "unable to determine resource")
```

File: pkg/server/handler.go [hunk 3]
```diff
@@ -223,14 +224,15 @@ func WithWorkspaceProjection(apiHandler http.Handler, shardVirtualWorkspaceURL *
...
+			logger = logger.WithValues("from", path.Join(cluster.Name.Path(), req.URL.Path), "to", newPath)
+			logger.V(4).Info("rewriting path")
 			req.URL.Path = newPath
 			if shardVirtualWorkspaceURL != nil {
 				u := *req.URL // shallow copy
 				u.RawQuery = req.URL.RawQuery
 				u.Host = shardVirtualWorkspaceURL.Host
 
-				klog.V(4).Infof("Redirecting virtual workspace request to %s", u.String())
+				logger.WithValues("redirect", u.String()).V(4).Info("redirecting virtual workspace request")
```
Comment (ncdc) [inline pkg/server/handler.go:235]: Is it better to do logger.V(4).WithValues to let the V check happen first?

File: pkg/server/home_workspaces.go [hunk 4]
```diff
@@ -517,12 +527,19 @@ func tryToCreate(h *homeWorkspaceHandler, ctx context.Context, user kuser.Info,
...
+			Annotations: map[string]string{logicalcluster.AnnotationKey: parent.String()},
 		},
 		Spec: tenancyv1alpha1.ClusterWorkspaceSpec{
 			Type: tenancyv1alpha1.ClusterWorkspaceTypeReference{Path: tenancyv1alpha1.RootCluster.String(), Name: workspaceType},
 		},
 	}
+	logger := logging.WithObject(klog.FromContext(ctx).WithValues(
+		"action", "tryToCreate",
+		"parent", parent.String(),
+		"clusterName", logicalClusterName,
```
Comment (ncdc) [inline pkg/server/home_workspaces.go:540]: is clusterName needed or won't it come from ws?

File: pkg/server/home_workspaces.go [hunk 5]
```diff
@@ -533,9 +550,9 @@ func tryToCreate(h *homeWorkspaceHandler, ctx context.Context, user kuser.Info,
 		}
 	}
 
-	klog.V(4).InfoS("tryToCreate: creating workspace", "parent", parent.String(), "workspace", ws.Name)
+	logger.V(4).Info("creating workspace")
 	err := h.kcp.createClusterWorkspace(ctx, parent, ws)
-	klog.V(4).InfoS("tryToCreate: creating workspace result", "parent", parent.String(), "workspace", ws.Name, "err", err)
+	logger.V(4).Info("error creating workspace result", "err", err)
```
Comment (ncdc) [inline pkg/server/home_workspaces.go:555]:
```suggestion
	logger.V(4).Info("create workspace result", "err", err)
```

File: pkg/server/home_workspaces_test.go [hunk 6]
```diff
@@ -873,7 +854,7 @@ func TestTryToCreate(t *testing.T) {
 				user := &kuser.DefaultInfo{
 					Name: testCase.userName,
 				}
-				retryAfterSeconds, createError := handler.tryToCreate(testCase.context, user, logicalcluster.New(testCase.workspaceName), tenancyv1alpha1.ClusterWorkspaceTypeName(testCase.workspaceType))
+				retryAfterSeconds, createError := handler.tryToCreate(context.TODO(), user, logicalcluster.New(testCase.workspaceName), tenancyv1alpha1.ClusterWorkspaceTypeName(testCase.workspaceType))
```
Comment (ncdc) [inline pkg/server/home_workspaces_test.go:857]: If the test cases don't need to specify a varying context, let's use `context.Background()` here

File: pkg/server/home_workspaces_test.go [hunk 7]
```diff
@@ -1245,7 +1225,7 @@ func TestCreateHomeWorkspaceRBACResources(t *testing.T) {
 					},
 				}.build()
 
-				createError := handler.createHomeWorkspaceRBACResources(testCase.context, testCase.userName, logicalcluster.New(testCase.workspaceName))
+				createError := handler.createHomeWorkspaceRBACResources(context.TODO(), testCase.userName, logicalcluster.New(testCase.workspaceName))
```
Comment (ncdc) [inline pkg/server/home_workspaces_test.go:1228]: If the test cases don't need to specify a varying context, let's use `context.Background()` here

File: pkg/server/options/controllers.go [hunk 8]
```diff
@@ -82,7 +82,7 @@ func (c *Controllers) Complete(rootDir string) error {
 		// use sa.key and auto-generate if not existing
 		c.SAController.ServiceAccountKeyFile = filepath.Join(rootDir, "sa.key")
 		if _, err := os.Stat(c.SAController.ServiceAccountKeyFile); os.IsNotExist(err) {
-			klog.Infof("Generating service account key file %s", c.SAController.ServiceAccountKeyFile)
+			klog.Background().WithValues("file", c.SAController.ServiceAccountKeyFile).Info("generating service account key file")
```
Comment (ncdc) [inline pkg/server/options/controllers.go:85]: If we wanted to, we could wire in a context from main.go down to here

File: pkg/server/options/options.go [hunk 9]
```diff
@@ -361,14 +361,15 @@ func mkdirRoot(dir string) error {
 	if dir == "" {
 		return errors.New("missing root directory configuration")
 	}
+	logger := klog.Background().WithValues("dir", dir)
```
Comment (ncdc) [inline pkg/server/options/options.go:364]: If we wanted to, we could wire in a context from main.go down to here

File: pkg/server/bootstrap/identity.go [hunk 10]
```diff
@@ -145,14 +154,21 @@ func wildcardIdentitiesResolver(ids *identities,
...
-			klog.V(2).Infof("APIExport %s|%s for group %q has identity %s", tenancyv1alpha1.RootCluster, name, group, apiExportIdentity)
+			logger.V(2).Info("APIExport has identity")
 		}
 		for gr, name := range groupResourceExportNames {
+			logger := logging.WithObject(logger, &v1alpha1.APIExport{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:        name,
+					Annotations: map[string]string{logging.APIVersionKey: tenancyv1alpha1.RootCluster.String()},
+				},
+			}).WithValues("gr", gr.String())
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply pkg/server/bootstrap/identity.go:171]: If I had to guess, @sttts would advise the opposite :)

File: pkg/server/home_workspaces.go (hunk 4 above)
Comment (stevekuznetsov) [reply pkg/server/home_workspaces.go:540]: I don't know, this flow was hard to follow. I was hoping we could clean it up after if it was to add redundant information.

File: pkg/server/options/options.go (hunk 9 above)
Comment (stevekuznetsov) [reply pkg/server/options/options.go:364]: Didn't seem like it would really benefit us at all, so I didn't

File: pkg/logging/constants.go [hunk 11]
```diff
@@ -92,3 +94,31 @@ func FromPrefix(prefix string, obj Object) []interface{} {
...
 	}
 }
+
+// WithUser adds user identifiers to the logger.
+func WithUser(logger logr.Logger, u user.Info) logr.Logger {
+	return logger.WithValues(FromUser(u)...)
+}
+
+// FromUser provides the structured logging fields that identify a user, prefixing with 'user'.
+func FromUser(u user.Info) []interface{} {
```
Comment (p0lyn0mial) [inline pkg/logging/constants.go:104]: does it have to be exported ?

File: pkg/logging/constants.go [hunk 12]
```diff
@@ -92,3 +94,31 @@ func FromPrefix(prefix string, obj Object) []interface{} {
...
+	return FromUserPrefix(u, "user")
+}
+
+// FromUserPrefix provides the structured logging fields that identify a user, allowing any prefix.
+func FromUserPrefix(u user.Info, prefix string) []interface{} {
+	return []interface{}{
+		prefix + ".name", u.GetName(),
```
Comment (p0lyn0mial) [inline pkg/logging/constants.go:111]: why not to use camelCase? (like WithCluster)

File: pkg/logging/constants.go [hunk 13]
```diff
@@ -92,3 +94,31 @@ func FromPrefix(prefix string, obj Object) []interface{} {
...
+		prefix + ".uid", u.GetUID(),
```
Comment (p0lyn0mial) [inline pkg/logging/constants.go:112]: I think that `UID`s can get very long, maybe we could just log some portion of it?

File: pkg/server/apiextensions.go [hunk 14]
```diff
@@ -69,10 +70,12 @@ var _ kcp.ClusterAwareCRDLister = &apiBindingAwareCRDLister{}
 // List lists all CustomResourceDefinitions that come in via APIBindings as well as all in the current
 // logical cluster retrieved from the context.
 func (c *apiBindingAwareCRDLister) List(ctx context.Context, selector labels.Selector) ([]*apiextensionsv1.CustomResourceDefinition, error) {
+	logger := klog.FromContext(ctx)
 	clusterName, err := request.ClusterNameFrom(ctx)
 	if err != nil {
 		return nil, err
 	}
+	logger = logger.WithValues("workspace", clusterName.String())
```
Comment (p0lyn0mial) [inline pkg/server/apiextensions.go:78]: we don't want to use `WithCluster` here?

File: pkg/logging/constants.go (hunk 11 above)
Comment (stevekuznetsov) [reply pkg/logging/constants.go:104]: Can't hurt, can it?

File: pkg/logging/constants.go (hunk 12 above)
Comment (stevekuznetsov) [reply pkg/logging/constants.go:111]: We've been using the period delimiter for fields of one object - we can update `WithCluster` to do that, too, no real strong feeling there. That seemed like more of a holder object for semi-unrelated data (e.g. cluster name requested and whether or not it's a partial metadata request, as opposed to e.g. the name/namespace of an object or name/uid of a user)

File: pkg/logging/constants.go (hunk 11 above)
Comment (p0lyn0mial) [reply pkg/logging/constants.go:104]: :)

**Comment by p0lyn0mial** [comment]
/lgtm

we can always change something in a separate PR
